example/FileSystem.tla

```diff
@@ -0,0 +1,138 @@
+------------------- MODULE FileSystem -----------------------
+\* A small example spec of some file system operations.
+\* This spec is meant for use in the `app.py` in the same directory,
+\* which demonstrates use of the chai client for interacting with
+\* the Apalache server.
+
+EXTENDS Apalache, Variants, Sequences, Integers, FiniteSets
+
+
+\*********\
+\* STATE *\
+\*********\
+
+CONSTANT
+    \* The names that can be used as path segments.
+    \* @type: Set(Str);
+    Names
+
+VARIABLES
+    \* The set of all files in the system.
+    \* @type: Set($file);
+    files,
+    \* The commend issued.
+    \* @type: $cmd;
+    cmd
+
+
+\**********\
+\* DOMAIN *\
+\**********\
+
+\** Representation of files, identified by their kind and their path.
+\*
+\* @typeAlias: path = Seq(Str);
+\* @typeAlias: file = File($path) | Dir($path);
+\* @type: $path => $file;
+File(path) == Variant("File", path)
+\* @type: $path => $file;
+Dir(path) == Variant("Dir", path)
+
+IsDir(f) == "Dir" = VariantTag(f)
+IsFile(f) == "File" = VariantTag(f)
+
+\* @type: $file => $path;
+FilePath(f) == VariantGetOrElse("Dir", f, VariantGetUnsafe("File", f))
+
+
+\** The children of a file (i.e., the contents of a directory)
+\*
+\* NOTE: The `Children` relation is reflexive: each file is its own child.
+\*
+\* @type: $file => Set($file);
+Children(p) ==
+  { child \in files: FilePath(p) = SubSeq(FilePath(child), 1, Len(FilePath(p))) }
+
+\* @type: Seq($file);
+NoFile == <<>>
+
+\** Representation of commands that operate on the file system.
+\*
+\* A $cmd is a triple holding:
+\*
+\* - the command name
+\* - the path(s) it operates on
+\* - the command's output (if any)
+\*
+\* @typeAlias: cmd = << Str, Seq($file), Set($file) >> ;
+\* @type: $file => $cmd;
+TouchCmd(p) == << "Touch", <<p>>, {} >>
+\* @type: $file => $cmd;
+MkDirCmd(p) == << "MkDir", <<p>>, {} >>
+\* @type: $file => $cmd;
+LsCmd(p)    == << "Ls", <<p>>, Children(p) >>
+\* @type: () => $cmd;
+None == << "None", NoFile, {} >>
+
+
+\***********\
+\* ACTIONS *\
+\***********\
+
+\* @type: Bool;
+Touch ==
+  \E d \in files:
+    /\ IsDir(d)
+    /\ \E n \in Names:
+      \* @type: $file;
+      LET newFile == File(Append(FilePath(d), n)) IN
+      /\ files' = { newFile } \union files
+      /\ cmd' = TouchCmd(newFile)
```

Comment:
Not sure if you want to consider this or not, but shouldn't this also include a check for whether a file with that name, in that path already exists? Maybe have an error variable to track whether the last operation succeeds?
I'd make `Touch` take `d` and `name` as parameters too, feels more intuitive:
```tla
\* @type: ($file, Str) => Bool;
Touch(d, name) ==
    /\ IsDir(d)
    /\ LET 
       \* @type: $file;
       newFile == File(Append(FilePath(d), n)) 
       IN
       /\ error' = newFile \in files
       /\ files' = { newFile } \union files \* If error' = true this is a no-op
       /\ cmd' = TouchCmd(newFile)
```
`MkDir` would be similar, then next would be
```tla
\E d \in files:
  \/ \E n \in Names:
    \/ Touch(d, n)
    \/ MkDir(d,n)
  \/ Ls(d)
```

Reply:
Fixed, thanks! b5c28f8ffefc665f92d0a413becf3a04bd905058